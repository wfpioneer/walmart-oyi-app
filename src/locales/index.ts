import * as RNLocalize from 'react-native-localize';
import I18n from 'i18n-js';
// @ts-ignore
import moment from 'moment';
import 'moment/locale/es';
import en from './en';
import es from './es';
import zh from './zh';

export const setI18nConfig: () => void = () => {
  // fallback if no available language fits
  const fallback = { languageTag: 'en', isRTL: false };

  const { languageTag } = RNLocalize.findBestAvailableLanguage(['en', 'es', 'zh'])
    || fallback;

  // set i18n-js config
  I18n.translations = { en, es, zh };
  I18n.locale = languageTag;

  // the following is to set up moment to work with spanish localization
  moment.locale(languageTag);
};

export const setLanguage = (language: string) => {
  if (language !== 'en' && language !== 'es' && language !== 'zh') {
    throw new Error('Unsupported language');
  }

  I18n.locale = language;
  moment.locale(language);
};

export const strings: (
  tag: string,
  options?: I18n.TranslateOptions | undefined,
) => string = (tag: string, options: I18n.TranslateOptions | undefined = {}) => I18n.t(tag, options);

export const currencies: (
  value: number,
  options?: I18n.ToCurrencyOptions | undefined
) => string = (value: number, options: I18n.ToCurrencyOptions | undefined = {}) => I18n.toCurrency(value, options);

export const numbers: (
  value: number,
  options?: I18n.ToNumberOptions | undefined
) => string = (value: number, options: I18n.ToNumberOptions | undefined = {}) => I18n.toNumber(value, options);
