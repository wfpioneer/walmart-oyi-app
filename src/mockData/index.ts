
import itemDetails from './getItemDetails';


export const getMockItemDetails = (upc: string) => {
  // @ts-ignore
  return itemDetails[upc] || {};
};
