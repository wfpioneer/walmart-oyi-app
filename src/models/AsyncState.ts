export interface AsyncState{
  isWaiting: boolean;
  value: any;
  error: any;
  result: any;
}
