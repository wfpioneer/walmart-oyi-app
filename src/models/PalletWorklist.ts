export type PalletWorklistType = 'MP';

export interface MissingPalletWorklistItemI {
  palletId: number;
  lastKnownPalletLocationId: number;
  lastKnownPalletLocationName: string;
  createUserId: string;
  createTs: string;
  palletDeleted: boolean;
  completed: boolean;
  completedTs?: string;
  completedUserId?: string;
  itemCount?: number;
  sectionID?: number;
}

// eslint-disable-next-line no-shadow
export enum Tabs {
  TODO='ToDo',
  COMPLETED='Completed'
}
