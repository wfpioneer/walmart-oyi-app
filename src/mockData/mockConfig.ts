import { Configurations, area } from '../models/User';

export const mockAreas: area[] = [
  {
    area: 'CENTER',
    categories: [5, 7, 9, 11, 13, 16]
  },
  {
    area: 'GM',
    categories: [21, 23, 24, 28, 31]
  },
  {
    area: 'Fresh',
    categories: [41, 42, 43, 45, 47, 48]
  }
];

export const mockConfig: Configurations = {
  locationManagement: false,
  locationManagementEdit: false,
  palletManagement: false,
  settingsTool: false,
  printingUpdate: false,
  binning: false,
  palletExpiration: false,
  backupCategories: '',
  picking: false,
  areas: mockAreas
};
