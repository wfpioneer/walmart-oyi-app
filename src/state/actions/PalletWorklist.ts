import { Tabs } from '../../models/PalletWorklist';

export const SET_SELECTED_TAB = 'PALLET_WORKLIST/SET_SELECTED_TAB';
export const SET_SELECTED_WORKLIST_PALLET_ID = 'PALLET_WORKLIST/SET_SELECTED_WORKLIST_PALLET_ID';
export const CLEAR_SELECTED_PALLET_WORKLIST_ID = 'PALLET_WORKLIST/CLEAR_SELECTED_PALLET_WORKLIST_ID';

export const setSelectedTab = (tab: Tabs) => ({
  type: SET_SELECTED_TAB,
  payload: tab
} as const);

export const setSelectedWorklistPalletId = (id: string) => ({
  type: SET_SELECTED_WORKLIST_PALLET_ID,
  payload: id
} as const);

export const clearSelectedWorklistPalletId = () => ({
  type: CLEAR_SELECTED_PALLET_WORKLIST_ID
} as const);

export type Actions =
  | ReturnType<typeof setSelectedTab>
  | ReturnType<typeof setSelectedWorklistPalletId>
  | ReturnType<typeof clearSelectedWorklistPalletId>
