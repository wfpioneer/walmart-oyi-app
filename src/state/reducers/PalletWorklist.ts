import { Tabs } from '../../models/PalletWorklist';
import {
  Actions,
  CLEAR_SELECTED_PALLET_WORKLIST_ID,
  SET_SELECTED_TAB,
  SET_SELECTED_WORKLIST_PALLET_ID
} from '../actions/PalletWorklist';

export interface PalletWorklistState {
   selectedTab: Tabs;
   selectedWorklistPalletId: string
  }

export const initialState: PalletWorklistState = {
  selectedTab: Tabs.TODO,
  selectedWorklistPalletId: ''
};

export const PalletWorklist = (state = initialState, action: Actions) => {
  switch (action.type) {
    case SET_SELECTED_WORKLIST_PALLET_ID:
      return {
        ...state,
        selectedWorklistPalletId: action.payload
      };
    case SET_SELECTED_TAB:
      return {
        ...state,
        selectedTab: action.payload
      };
    case CLEAR_SELECTED_PALLET_WORKLIST_ID:
      return {
        ...state,
        selectedWorklistPalletId: ''
      };
    default:
      return state;
  }
};
