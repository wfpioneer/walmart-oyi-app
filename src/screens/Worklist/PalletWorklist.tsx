import React, {
  Dispatch, EffectCallback, useEffect, useState
} from 'react';
import { FlatList, Text, View } from 'react-native';
import { useDispatch } from 'react-redux';
import Toast from 'react-native-toast-message';
import { NavigationProp, useNavigation } from '@react-navigation/native';
import MissingPalletWorklistCard from '../../components/MissingPalletWorklistCard/MissingPalletWorklistCard';
import { strings } from '../../locales';
import { MissingPalletWorklistItemI } from '../../models/WorklistItem';
import { CustomModalComponent } from '../Modal/Modal';
import { styles } from './PalletWorklist.style';
import Button from '../../components/buttons/Button';
import COLOR from '../../themes/Color';
import { clearPallet } from '../../state/actions/saga';
import { useTypedSelector } from '../../state/reducers/RootReducer';
import { AsyncState } from '../../models/AsyncState';
import {
  hideActivityModal,
  showActivityModal
} from '../../state/actions/Modal';
import { CLEAR_PALLET } from '../../state/actions/asyncAPI';

interface PalletWorkListProps {
  palletWorklist: MissingPalletWorklistItemI[];
  displayConfirmation: boolean;
  setDisplayConfirmation: React.Dispatch<React.SetStateAction<boolean>>;
  dispatch: Dispatch<any>;
  clearPalletAPI: AsyncState;
  navigation: NavigationProp<any>;
  useEffectHook: (effect: EffectCallback, deps?: ReadonlyArray<any>) => void;
}

export const clearPalletAPIHook = (
  clearPalletApi: AsyncState,
  palletId: string,
  navigation: NavigationProp<any>,
  dispatch: Dispatch<any>,
  setDisplayConfirmation: React.Dispatch<React.SetStateAction<boolean>>
): void => {
  if (navigation.isFocused()) {
    if (!clearPalletApi.isWaiting) {
      // Success
      if (clearPalletApi.result) {
        dispatch(hideActivityModal());
        setDisplayConfirmation(false);
        dispatch({ type: CLEAR_PALLET.RESET });

        // TODO REFRESH/CALL MISSING PALLET WORKLIST API

        Toast.show({
          type: 'success',
          text1: strings('PALLET.CLEAR_PALLET_SUCCESS', { palletId }),
          position: 'bottom'
        });
      }

      // Failure
      if (clearPalletApi.error) {
        dispatch(hideActivityModal());
        setDisplayConfirmation(false);
        dispatch({ type: CLEAR_PALLET.RESET });
        Toast.show({
          type: 'error',
          text1: strings('PALLET.CLEAR_PALLET_ERROR'),
          text2: strings('GENERICS.TRY_AGAIN'),
          position: 'bottom'
        });
      }
    } else {
      dispatch(showActivityModal());
    }
  }
};

export const PalletWorklistScreen = (props: PalletWorkListProps) => {
  const {
    clearPalletAPI,
    displayConfirmation,
    dispatch,
    palletWorklist,
    setDisplayConfirmation,
    navigation,
    useEffectHook
  } = props;
  let deletePalletId = '';

  useEffectHook(() => clearPalletAPIHook(
    clearPalletAPI,
    deletePalletId,
    navigation,
    dispatch,
    setDisplayConfirmation
  ),
  [clearPalletAPI]);

  const onDeletePress = () => {
    dispatch(clearPallet({ palletId: deletePalletId }));
  };

  const handleDeleteClick = (palletId: string) => {
    setDisplayConfirmation(true);
    deletePalletId = palletId;
  };

  const handleAddLocationClick = () => navigation.navigate('ScanPallet');

  return (
    <View style={styles.container}>
      <CustomModalComponent
        isVisible={displayConfirmation}
        modalType="FormHeader"
        onClose={() => {}}
      >
        <View style={styles.delHeader}>
          <Text style={styles.message}>
            {strings('MISSING_PALLET_WORKLIST.DELETE_PALLET')}
          </Text>
        </View>
        <Text style={styles.delText}>
          {strings('MISSING_PALLET_WORKLIST.DELETE_PALLET_CONFIRMATION')}
        </Text>
        <View style={styles.buttonContainer}>
          <Button
            style={styles.delButton}
            title={strings('GENERICS.CANCEL')}
            backgroundColor={COLOR.MAIN_THEME_COLOR}
            onPress={() => setDisplayConfirmation(false)}
          />
          <Button
            style={styles.delButton}
            title={strings('GENERICS.OK')}
            backgroundColor={COLOR.TRACKER_RED}
            onPress={() => onDeletePress()}
          />
        </View>
      </CustomModalComponent>
      <FlatList
        data={palletWorklist}
        keyExtractor={(item: MissingPalletWorklistItemI, index: number) => item.palletId + index.toString()}
        renderItem={({ item }) => (
          <MissingPalletWorklistCard
            palletId={item.palletId}
            lastLocation={item.lastKnownLocationName}
            reportedBy={item.createId}
            reportedDate={item.createTS}
            expanded={true} // TODO Toggle for a single Pallet WorkList Item
            addCallback={handleAddLocationClick}
            deleteCallback={() => handleDeleteClick(item.palletId.toString())}
            navigateCallback={() => {}}
          />
        )}
        onRefresh={null}
        refreshing={undefined}
        style={styles.list}
      />
    </View>
  );
};

export const PalletWorkList = () => {
  const [displayConfirmation, setDisplayConfirmation] = useState(false);
  const clearPalletAPI = useTypedSelector(state => state.async.clearPallet);
  const dispatch = useDispatch();
  const navigation = useNavigation();
  const mockMPWorklist: MissingPalletWorklistItemI[] = [
    {
      createId: '11',
      createTS: '26/06/2022',
      lastKnownLocationId: 1,
      lastKnownLocationName: 'A1-1',
      palletDeleted: false,
      palletId: 7988,
      worklistType: 'MP',
      completed: undefined,
      completedId: undefined,
      completedTS: undefined
    }
  ];
  return (
    <PalletWorklistScreen
      palletWorklist={mockMPWorklist}
      displayConfirmation={displayConfirmation}
      setDisplayConfirmation={setDisplayConfirmation}
      dispatch={dispatch}
      clearPalletAPI={clearPalletAPI}
      navigation={navigation}
      useEffectHook={useEffect}
    />
  );
};
