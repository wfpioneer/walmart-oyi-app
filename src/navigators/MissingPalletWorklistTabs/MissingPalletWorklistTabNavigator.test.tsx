import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {
  NavigationProp,
  RouteProp
} from '@react-navigation/native';
import Toast from 'react-native-toast-message';
import { strings } from '../../locales';
import {
  MissingPalletWorklistTabNavigator, getPalletConfigHook, getPalletDetailsApiHook
} from './MissingPalletWorklistTabNavigator';
import { mockConfig } from '../../mockData/mockConfig';
import { AsyncState } from '../../models/AsyncState';
import { Tabs } from '../../models/PalletWorklist';
import { SNACKBAR_TIMEOUT } from '../../utils/global';

let navigationProp: NavigationProp<any>;
const routeProp: RouteProp<any, string> = {
  key: '',
  name: 'MissingPalletWorklistTabs'
};

describe('MissingPalletWorklist Navigator', () => {
  const defaultAsyncState: AsyncState = {
    isWaiting: false,
    value: null,
    error: null,
    result: null
  };
  it('Renders the Missing Pallet Worklist navigator component', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <MissingPalletWorklistTabNavigator
        useFocusEffectHook={jest.fn()}
        dispatch={jest.fn()}
        navigation={navigationProp}
        route={routeProp}
        validateSessionCall={jest.fn()}
        useCallbackHook={jest.fn()}
        useEffectHook={jest.fn()}
        perishableCategories={[49, 20]}
        selectedTab={Tabs.TODO}
        getPalletDetailsApi={defaultAsyncState}
        getPalletConfigApi={defaultAsyncState}
        userConfig={mockConfig}
        palletClicked={false}
        setPalletClicked={jest.fn()}
        getPalletDetailsComplete={false}
        setGetPalletDetailsComplete={jest.fn()}
        palletConfigComplete={false}
        setPalletConfigComplete={jest.fn()}
      />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  describe('MissingPalletWorklistTabNavigator externalized function tests', () => {
    const mockDispatch = jest.fn();
    afterEach(() => {
      jest.clearAllMocks();
    });
    const mockSetGetPalletDetailsComplete = jest.fn();
    const mockSetPalletClicked = jest.fn();

    it('Tests getPalletDetailsApiHook on success', () => {
      const successApi: AsyncState = {
        ...defaultAsyncState,
        result: {
          status: 200,
          data: {
            pallets: [
              {
                id: 1,
                createDate: 'today',
                expirationDate: 'tomorrow',
                items: []
              }
            ]
          }
        }
      };

      getPalletDetailsApiHook(
        successApi, true, mockSetPalletClicked, mockDispatch, mockSetGetPalletDetailsComplete
      );
      expect(mockDispatch).toBeCalledTimes(2);
      expect(Toast.show).toBeCalledTimes(0);
      expect(mockSetPalletClicked).toHaveBeenCalledWith(false);
      expect(mockSetGetPalletDetailsComplete).toHaveBeenCalledWith(true);
    });

    it('Tests getPalletDetailsApiHook with status 204', () => {
      const apiResult: AsyncState = {
        ...defaultAsyncState,
        result: {
          status: 204,
          data: {}
        }
      };

      getPalletDetailsApiHook(
        apiResult, false, mockSetPalletClicked, mockDispatch, mockSetGetPalletDetailsComplete
      );
      expect(Toast.show).toHaveBeenCalledWith({
        type: 'error',
        text1: strings('LOCATION.PALLET_NOT_FOUND'),
        visibilityTime: SNACKBAR_TIMEOUT,
        position: 'bottom'
      });
    });

    it('Tests getPalletDetailsApiHook on fail', () => {
      const failApi: AsyncState = { ...defaultAsyncState, error: {} };
      getPalletDetailsApiHook(
        failApi, false, mockSetPalletClicked, mockDispatch, mockSetGetPalletDetailsComplete
      );
      expect(Toast.show).toHaveBeenCalledWith({
        type: 'error',
        text1: strings('PALLET.PALLET_DETAILS_ERROR'),
        text2: strings('GENERICS.TRY_AGAIN'),
        visibilityTime: SNACKBAR_TIMEOUT,
        position: 'bottom'
      });
    });
    it('test getPalletConfigHook', async () => {
      const dispatch = jest.fn();
      const setConfigComplete = jest.fn();
      const successAsyncState = {
        ...defaultAsyncState,
        result: {
          status: 200,
          data: {
            perishableCategories: [1, 8, 35, 36, 40, 41, 43, 45, 46, 47, 49, 51, 52, 53, 54, 55, 58]
          }
        }
      };
      const failureAsyncState = {
        ...defaultAsyncState,
        error: 'test'
      };

      getPalletConfigHook(successAsyncState, dispatch, setConfigComplete, '1, 8');
      expect(dispatch).toBeCalledTimes(2);
      expect(setConfigComplete).toBeCalledTimes(1);

      dispatch.mockReset();
      setConfigComplete.mockReset();
      getPalletConfigHook(failureAsyncState, dispatch, setConfigComplete, '1, 8');
      expect(dispatch).toBeCalledTimes(2);
      expect(setConfigComplete).toBeCalledTimes(1);
    });
  });
});
