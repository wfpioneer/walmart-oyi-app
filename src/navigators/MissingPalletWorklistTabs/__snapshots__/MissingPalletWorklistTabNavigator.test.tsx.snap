// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissingPalletWorklist Navigator Renders the Missing Pallet Worklist navigator component 1`] = `
<MaterialTopTabNavigator
  initialRouteName="ToDo"
  screenOptions={
    Object {
      "tabBarActiveTintColor": "#ffffffff",
      "tabBarIndicatorStyle": Object {
        "backgroundColor": "#ffffffff",
      },
      "tabBarStyle": Object {
        "backgroundColor": "#0071e9",
      },
    }
  }
>
  <Screen
    listeners={
      Object {
        "focus": [Function],
      }
    }
    name="[missing \\"en.WORKLIST.TODO\\" translation]"
  >
    [Function]
  </Screen>
  <Screen
    listeners={
      Object {
        "focus": [Function],
      }
    }
    name="[missing \\"en.WORKLIST.COMPLETED\\" translation]"
  >
    [Function]
  </Screen>
</MaterialTopTabNavigator>
`;
